core/src/main/java/com/digitalpebble/stormcrawler/protocol/HttpRobotRulesParser.java

```diff
@@ -165,16 +168,20 @@ public BaseRobotRules getRobotRulesSet(Protocol http, URL url) {
                             response.getContent() != null ? response.getContent().length : 0);
                 }
             }
-            if (code == 200) // found rules: parse them
-            {
+
+            // Parsing found rules; by default, all robots are forbidden (RFC 9309)
+            robotRules = FORBID_ALL_RULES;
```

Comment:
The default for all HTTP response codes not handled explicitly is now forbid-all instead of allow-all. Are there codes except 200, 403, 5xx, redirects which need a special treatment? Maybe not.

Reply:
> Maybe not.

Sorry, I was wrong. In case of a HTTP 404 (robots.txt not found), crawling is allowed.

In doubt, maybe do not swap the defaults? "In the wild" the crawler may see any HTTP status code.

Reply:
Oh yes, how could I forget about status code 404?! After 200 and redirects, it is probably what the crawler encounters most often when fetching the robots.txt.
In the RFC 9309, it is written that for all status codes in the range of 400-499 the robots file may be considered as "unavailable" and therefore allow-all. This is also true for 300-399 in case the crawler followed at least five consecutive redirects.

This implementation might be better:

```
// TODO: Follow up to five redirections (at the moment, it is only one)

if (200 <= code && code <= 299) {
	String ct = response.getMetadata().getFirstValue(HttpHeaders.CONTENT_TYPE);
	robotRules = parseRules(url.toString(), response.getContent(), ct, agentNames);
} else if (300 <= code && code <= 499) {
	robotRules = EMPTY_RULES; // allow all
	if (code == 403 && !allowForbidden) {
		robotRules = FORBID_ALL_RULES; // forbid all
	}
	// E.g. Google handles Too many requests similar to a server error
	// https://support.google.com/webmasters/answer/9679690#robots_details
	if (code == 429) {
		robotRules = FORBID_ALL_RULES; // forbid all
	}
} else if (500 <= code) {
	cacheRule = false;
	robotRules = FORBID_ALL_RULES; // forbid all
	if (allow5xx) {
		robotRules = EMPTY_RULES; // allow all
	}
} else {
	robotRules = EMPTY_RULES; // allow all
}
```
I'm not sure whether e.g. `if (200 <= code && code <= 299)` or `if (code == 200)` is more correct?
In case of `if (300 <= code && code <= 499)` and `if (code == 301 || code == 302 || code == 403 || code == 404 || code == 410)` it is not important because both options would result in allow-all

Reply:
> After 200 and redirects, it is probably what the crawler encounters most often when fetching the robots.txt

Yes. In the long tail the crawler gets almost every possible HTTP status code, see the attached [robotstxt-cc-main-2023-14.txt](https://github.com/DigitalPebble/storm-crawler/files/11247290/robotstxt-cc-main-2023-14.txt).

That's why I'd be careful about `if (200 <= code && code <= 299)` or `else if (500 <= code)`.

> In the RFC 9309, it is written that for all status codes in the range of 400-499 the robots file may be considered as "unavailable" and therefore allow-all.

Yes. The 403 (and 401) handling is defined in the "original" [norobots-RFC](https://www.robotstxt.org/norobots-rfc.txt). So, there are some differences - you mentioned the 429 handled by Google's crawler.

> // TODO: Follow up to five redirections (at the moment, it is only one)

Good catch!

Maybe keep this in a separate issue / PR to limit the amount of changes for an easy review. A unit test would be also useful to ensure that no change accidentally breaks something.

Reply:
> Yes. In the long tail the crawler gets almost every possible HTTP status code, see the attached [robotstxt-cc-main-2023-14.txt](https://github.com/DigitalPebble/storm-crawler/files/11247290/robotstxt-cc-main-2023-14.txt).

Wow, impressing numbers. Thanks for sharing. While crawling, you'll always have to be prepared to the unexpected, this was not clear to me in this extent.
As a consequence, the condition `else if (500 <= code)` is definitely wrong and has to be substituted with `else if (500 <= code && code <= 599)` in order to implement more precisely what is written in the RFC. Otherwise, the crawler would restrict itself more than necessary.

Besides that, there are probably two approaches: First, stick with the old way to go and parse the Robots.txt file only for Status code 200. In case the Status code is weird and unexpected, such as 204, 711 or 1, the crawler assumes the default ALLOW_ALL.
Or second, parse the Robots.txt file for any Status code except 300-599. This is probably a bit closer to the concrete wording in the RFC, in which no status code range is mentioned, as it only speaks of a "successful download".
```
2.3.1.1. Successful Access
If the crawler successfully downloads the robots.txt file, the
crawler MUST follow the parseable rules.
```
I think, in this case, the crawler wouldn't restrict itself too much either, as the parseContent() method by crawlercommons.robots.SimpleRobotRulesParser, which is furtheron called, also assumes ALLOW_ALL as default for an empty or unparsable Robots.txt file etc.